pep-0724.rst

```diff
@@ -0,0 +1,235 @@
+PEP: 724
+Title: Stricter Type Guards
+Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <erictr at microsoft.com>, Erik De Bonte <erikd at microsoft.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 28-Jul-2023
+Python-Version: 3.13
+Post-History: 
+Resolution: 
+
+
+Abstract
+========
+
+:pep:`647` introduced the concept of ``TypeGuard`` functions which return true
+if their input parameter matches their target type. For example, a function that
+returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's
+input parameter is a ``str``. This allows type checkers to narrow types in this
+positive case.
+
+This PEP further refines :pep:`647` by allowing type checkers to also narrow types
+when a ``TypeGuard`` function returns false.
+
+Motivation
+==========
+
+``TypeGuards`` are used throughout Python libraries to allow a type checker to
+narrow the type of something when the ``TypeGuard`` returns true.
+
+However, in the ``else`` clause, :pep:`647` didn't prescribe what the type might
+be:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type here is not narrowed. It is still 'str | int' 
+
+
+This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype
```

Comment:
The term "subtype" is probably not sufficient here. Subtype relationships are well-defined in the Python type system as long as there is no `Any` involved. As soon as you introduce an `Any`, you can no longer talk about subtyping. Instead, you need to talk about "type consistency" or "consistent types". See PEP 483 for a discussion of how this differs.

For this spec, I should use "consistent types" rather than "subtypes" as the test. That's how I've coded the current prototype.

Reply:
Didn't realize there was a 'is-consistent' term that deals with Any. I'll update this text and add a footnote about PEP 483